### ISSUE TITLE
Can't WaitForResourceAsync KnownResourceStates.Finished

### ISSUE THREAD

**Issue by eerhardt** [dotnet/aspire#4878]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

It's not possible to get notification when a resource has been stopped/finished when after `await app.StopAsync();`.

I'm trying to write a test that uses a Docker Volume, and when the test is done, it deletes the volume. However, the volume is still in use even after `await app.StopAsync();` returns, which means the container using the volume is still running.

I'd like a way to get notified when the resource/container is finished so I can delete the volume.

One reason this isn't possible today is that WaitForResourceAsync cancels once the app is stopping here:
https://github.com/dotnet/aspire/blob/681f2e754dc849f96abafcd2829783e69155abf1/src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs#L89-L93

Once the app is stopping, it cancels this loop, which means we won't get more notifications about the resource.

### Expected Behavior

I'd like to get a notification when the container is finished.

### Steps To Reproduce

```csharp
await using var app = await appHost.BuildAsync();
var resourceNotificationService = app.Services.GetRequiredService<ResourceNotificationService>();
await app.StartAsync();

await app.StopAsync();

await resourceNotificationService.WaitForResourceAsync("webfrontend", KnownResourceStates.Finished).WaitAsync(TimeSpan.FromSeconds(30));
```

### Anything else?

cc @DamianEdwards @karolz-ms @mitchdenny

**Comment by karolz-ms** [comment]
@eerhardt and I chatted. We think the main issue is that as soon as `app.StopAsync()` is called, the resource service and ApplicationExecutor stops listening to resource changes. So even if `DcpOptions.DeleteResourcesOnShutdown` is set, there is no way for the Aspire model to learn that the container (or any resource for that matter) has changed state, or was deleted.

It would probably make sense to change this behavior and differentiate between stopping the application (which would stop application executables and containers, but preserve the Aspire model and its ability to get state updates) vs disposing of the application (which would dispose of the model, stop DCP orchestrator etc.)

**Comment by davidfowl** [comment]
@DamianEdwards this is the code you added to unhook the notification service on shutdown automatically. We'd need to allow listening to changes past the stopping of the application (but before disposal)

**Comment by DamianEdwards** [comment]
> We'd need to allow listening to changes past the stopping of the application (but before disposal)

Is there an existing signal for that or would we need to do something custom?

**Comment by davidfowl** [comment]
I think shutting down in dispose would be better (implement IDisposable or IAsyncDisposable)

**Comment by karolz-ms** [comment]
This issue covers 3 different use cases: stopping a resource vs deleting a resource vs disposing of the application. We should recognize that they are different and are useful for different use cases. We should be clear about what use case we are talking about, what mechanism(s) we want to use for that use case(s), and what works as expected vs needs improvement. Some thoughts on the original use case that prompted @eerhardt to open this issue follow.

To me the right mechanism seems deleting the container that is using a volume. When the container is deleted, the app host should receive a notification and can proceed to delete the container volume that the container was using. This is all supported by the app orchestrator today and should work, provided the app model allows the app host code to do this (not 100% sure, some work may be necessary).

Alternatively the orchestrator could support marking the volume as "should be deleted when the app shuts down". That would be a new feature for the orchestrator. The test could then dispose the whole app and rely on automatic cleanup by app orchestrator.

Stopping a container in this scenario is not sufficient, because a stopped container is still holding a reference to the volume. That is just how Docker/Podman work.

Disposing the app host while listening to notifications is not going to work because (once https://github.com/dotnet/aspire/pull/7098 is merged) disposing the app triggers wholesale resource cleanup and orchestrator shutdown. The app orchestrator does not make any guarantees about the order in which resources are cleaned up, and it does not guarantee sending notifications about resources being cleaned up either. It just tries to get the machine to clean state as quickly as possible.

**Comment by DamianEdwards** [comment]
> I think shutting down in dispose would be better (implement IDisposable or IAsyncDisposable)

> Disposing the app host while listening to notifications is not going to work because (once https://github.com/dotnet/aspire/pull/7098 is merged) disposing the app triggers wholesale resource cleanup and orchestrator shutdown. The app orchestrator does not make any guarantees about the order in which resources are cleaned up, and it does not guarantee sending notifications about resources being cleaned up either. It just tries to get the machine to clean state as quickly as possible.

Regardless, it seems it would still make sense to change `ResourceNotificationService` to not kill the loop until it's disposed so that notifications sent after app host stopping but before disposal can be seen.

**Comment by karolz-ms** [comment]
> Regardless, it seems it would still make sense to change `ResourceNotificationService` to not kill the loop until it's disposed so that notifications sent after app host stopping but before disposal can be seen.

Agreed

**Comment by davidfowl** [comment]
It turns out when you let DCP shutdown it may not put resources in the Finished state. I observed Stopping and then nothing. This doesn't feel like a bug but we need to decide if this is the desired behavior.

https://github.com/dotnet/aspire/compare/davidfowl/stop-listeing-ondispose?expand=1

**Comment by DamianEdwards** [comment]
Yeah this is what I saw IIRC. If you wait a bit after calling `.Stop()` do more state changes come through? Or is it the case that once you ask DCP to shutdown (via `.Stop()`) it simply doesn't stick around to ensure it can pump the final resource states back to the app host? Would such a change in DCP be simplish to make? I appreciate that even if DCP sends those updates there's no guarantee the app host reads them from the notification stream before DCP shuts down the endpoint, unless we want to add more coordination to the shutdown process.

**Comment by DamianEdwards** [comment]
@davidfowl so if you wait after calling Stop then you DCP will stay around long enough to send you terminal states for the resources as they shutdown? Doesn't that suggest that the code @eerhardt shared originally should actually work now, which isn't what I was seeing:

```csharp
await app.StopAsync();

// The app host stays alive while we're waiting here so assuming the resource shuts down within 30 seconds we should get this update from DPC
await resourceNotificationService.WaitForResourceAsync("webfrontend", KnownResourceStates.Finished).WaitAsync(TimeSpan.FromSeconds(30));
```

**Comment by davidfowl** [comment]
Right this does not work even after these changes. I did this:

```C#
await app.StopAsync();

var resourceNotificationService = app.Services.GetRequiredService<ResourceNotificationService>();

await foreach (var re in resourceNotificationService.WatchAsync())
{
      Console.WriteLine($"Resource {re.Resource.Name} is in state {re.Snapshot.State!.Text}");
}
```

**Comment by DamianEdwards** [comment]
> Right this does not work even after these changes.

OK, so I'm a bit confused then. In [this comment](https://github.com/dotnet/aspire/issues/4878#issuecomment-2607797484) I was trying to ask why does this not work? Is it because we're not waiting long enough after calling `StopAsync`, or is it because DCP doesn't even try to stay alive long enough to send the final resource states after receiving a shutdown request? It sounds more like the latter, but I thought you were saying it was the former.

**Comment by davidfowl** [comment]
The latter. It seems like it reports stopping then dies, so I never see finished or exited.

**Comment by DamianEdwards** [comment]
OK that makes sense then.

@karolz-ms what are your thoughts on having DCP coordinate reporting of final resource status after a shutdown request is received such that it doesn't stop the update stream and/or shut itself down until resources are stopped?

**Comment by davidfowl** [comment]
Not in this change:

https://github.com/dotnet/aspire/issues/4878#issuecomment-2606418497

**Comment by davidfowl** [comment]
If it's easy we can do it in 9.1 but we don't have plans to take advantage of this in 9.1

**Comment by karolz-ms** [comment]
Synced up with @eerhardt offline. What @davidfowl (stop resources call + stop DCP call) seems to meet expectations for the scenario.

**Comment by karolz-ms** [comment]
We should re-triage, this might be necessary to prevent https://github.com/dotnet/aspire/issues/7215

**Comment by DamianEdwards** [comment]
Is the work left here to update `Aspire.Hosting` to issue the relevant two-phase commands to DCP appropriately, i.e. Stop() -> stop, Dispose() -> shutdown?

**Comment by davidfowl** [comment]
Yes we never merged https://github.com/dotnet/aspire/issues/4878#issuecomment-2606418497

**Comment by DamianEdwards** [comment]
@davidfowl OK so DCP got updates to ensure it can send final states now and we just need to merge your changes?

**Comment by davidfowl** [comment]
I assume we also need to call the new APIs in DCP yep.

**Comment by DamianEdwards** [comment]
OK, is this correctly assigned? @karolz-ms are you planning to do the work on the Aspire side for this?

**Comment by karolz-ms** [comment]
Yes, I am working on this as we speak. Specifically, on adding a new property to `DcpExecutor` that determines whether `StopAsync()` waits for resource cleanup to be complete or not. That will then be leveraged by test application builder. Should have a PR out for review sometime later today.

**Comment by karolz-ms** [comment]
Should be resolved with https://github.com/dotnet/aspire/pull/7446

**Comment by davidfowl** [comment]
I think we need a test to actually use this pattern to see if we need more changes to the app model codez.

**Comment by DamianEdwards** [comment]
> I think we need a test to actually use this pattern to see if we need more changes to the app model codez.

I can update the samples tests to do it. Is there are code snippet someone can point me at?

**Comment by eerhardt** [comment]
> I think we need a test to actually use this pattern to see if we need more changes to the app model codez.

Also https://github.com/dotnet/aspire/pull/7477 could use this as well.

cc @sebastienros
